File: golang/monotonic_stack/online_stock_span.go
package main

type StockSpanner struct {
	stack []struct {
		price int
		span  int
	}
}

func Constructor() StockSpanner {
	return StockSpanner{}
}

func (this *StockSpanner) Next(price int) int {
	span := 1

	for len(this.stack) > 0 && this.stack[len(this.stack)-1].price <= price {
		span += this.stack[len(this.stack)-1].span

		this.stack = this.stack[:len(this.stack)-1]
	}

	this.stack = append(this.stack, struct {
		price int
		span  int
	}{price, span})

	return span
}

File: java/src/main/java/com/sanedge/interview/monotonic_stack/OnlineStockSpan.java
package com.sanedge.interview.monotonic_stack;

import java.util.Stack;

public class OnlineStockSpan {
  private static class Stock {
    int price;
    int span;

    Stock(int price, int span) {
      this.price = price;
      this.span = span;
    }
  }

  private Stack<Stock> stack;

  public OnlineStockSpan() {
    stack = new Stack<>();
  }

  public int next(int price) {
    int span = 1;

    while (!stack.isEmpty() && stack.peek().price <= price) {
      span += stack.pop().span;
    }

    stack.push(new Stock(price, span));

    return span;
  }

}
